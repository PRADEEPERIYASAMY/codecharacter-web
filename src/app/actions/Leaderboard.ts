import * as LeaderboardInterfaces from 'app/types/Leaderboard';
import { action } from 'typesafe-actions';
export namespace LeaderboardActions {
  export enum Type {
    GET_PLAYERS_DATA = 'GET_PLAYERS_DATA',
    UPDATE_ERROR = 'UPDATE_ERROR',
    UPDATE_LEADERBOARD = 'UPDATE_LEADERBOARD',
  }

  export const MAX_RATING = 4000;

  export const updateLeaderboard = (leaderboardData: LeaderboardInterfaces.LeaderboardDataType) => {
    return action(Type.UPDATE_LEADERBOARD, leaderboardData);
  };

  export const updateError = (err: string) => {
    return action(Type.UPDATE_ERROR, err);
  };

  export const getPlayersData = () => {
    try {
      /* Implement async action using redux-saga for fetching data */
      /* Sample data */
      const data = {
        loading: false,
        players: [{ id: 1, name: 'Mutko', rating: 1, country: 'ru' }],
      };
      return updateLeaderboard(data);
    } catch (e) {
      return updateError(e);
    }
  };
}
