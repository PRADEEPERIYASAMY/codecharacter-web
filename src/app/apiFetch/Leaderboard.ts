import * as LeaderboardInterfaces from 'app/types/Leaderboard';
import { API_BASE_URL } from '../../config/config';

export const getLeaderboard = (body: LeaderboardInterfaces.GetLeaderboard) => {
  return fetch(`${API_BASE_URL}leaderboard/${body.pattern}/${body.start}/${body.end}`, {
    credentials: 'include',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'GET',
  })
    .then((response) => response.json())
    .then((data) => data)
    .catch((error) => {
      throw error;
    });
};
