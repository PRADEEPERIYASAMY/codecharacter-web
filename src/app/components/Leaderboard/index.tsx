import { LeaderboardElement } from 'app/components/Leaderboard/LeaderboardElement';
import * as styles from 'app/styles/Leaderboard.module.css';
import * as LeaderboardInterfaces from 'app/types/Leaderboard';
import classnames from 'classnames';
import * as React from 'react';
import { Col, Grid, Row } from 'react-bootstrap';

export class Leaderboard extends React.Component<LeaderboardInterfaces.Props, {}> {
  public render() {
    const { players } = this.props;
    return (
      <Grid fluid={true} className={classnames(styles.Leaderboard)}>
        <Row className="justify-content-between py-2 pl-3">
          <Col className="text-light font-weight-bold my-auto">LEADERBOARD</Col>
        </Row>
        <Row className={styles['leaderboard-wrap']}>
          {players.length ? (
            players.map((player, rank) => (
              <LeaderboardElement player={player} rank={rank} key={rank} />
            ))
          ) : (
            <div>Nothing to show</div>
          )}
        </Row>
      </Grid>
    );
  }
}
