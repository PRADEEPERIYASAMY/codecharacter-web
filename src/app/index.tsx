import { Dashboard } from 'app/components/Dashboard';
import * as React from 'react';
import { hot } from 'react-hot-loader';
import { Route, Switch } from 'react-router';

/* tslint:disable-next-line:variable-name */
export const App = hot(module)(() => (
  <Switch>
    <Route path="/" component={Dashboard} />
  </Switch>
));
