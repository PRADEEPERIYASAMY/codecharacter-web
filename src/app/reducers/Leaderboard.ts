import { LeaderboardActions } from 'app/actions';
import * as LeaderboardInterfaces from 'app/types/Leaderboard';

const leaderboardInitialState: LeaderboardInterfaces.LeaderboardStoreState = {
  loading: false,
  players: [],
};

export const leaderboardReducer = (
  state = leaderboardInitialState,
  action: LeaderboardInterfaces.LeaderboardStoreAction,
) => {
  switch (action.type) {
    case LeaderboardActions.Type.UPDATE_LOADING_STATUS: {
      return {
        ...state,
        loading: action.payload.loadingStatus,
      };
    }
    case LeaderboardActions.Type.UPDATE_LEADERBOARD: {
      let players = [];
      if (action.payload.type === LeaderboardActions.updateType.APPEND) {
        players = [...state.players, ...action.payload.leaderboardData];
      } else {
        players = action.payload.leaderboardData;
      }
      return {
        ...state,
        players,
        loading: false,
      };
    }
    case LeaderboardActions.Type.UPDATE_ERROR:
      return {
        ...state,
        loading: false,
      };
    default:
      return state;
  }
};
