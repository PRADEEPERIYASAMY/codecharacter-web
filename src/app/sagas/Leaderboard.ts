import { LeaderboardActions } from 'app/actions';
import * as LeaderboardFetch from 'app/apiFetch/Leaderboard';
import { resType } from 'app/types/sagas';
import { all, call, put, takeEvery } from 'redux-saga/effects';
import { ActionType } from 'typesafe-actions';

export function* getLeaderboard(action: ActionType<typeof LeaderboardActions.getLeaderboard>) {
  try {
    yield put(LeaderboardActions.updateLoadingStatus(true));

    const result = yield call(LeaderboardFetch.getLeaderboard, {
      end: action.payload.end,
      pattern: action.payload.pattern === '' ? '*' : action.payload.pattern,
      start: action.payload.start,
    });

    yield put(LeaderboardActions.updateError(result.error));

    // rank: number;
    // country: string;
    // username: string;
    // rating: number;
    // fullName: string;

    if (result.type === resType.ERROR) {
      yield put(
        LeaderboardActions.updateLeaderboard(
          // result.searchData,
          [
            {
              country: 'IN',
              fullName: 'FullName1',
              rank: 1,
              rating: 1,
              username: 'Username1',
            },
            {
              country: 'KE',
              fullName: 'FullName2',
              rank: 2,
              rating: 2,
              username: 'kenya',
            },
            {
              country: 'IN',
              fullName: 'FullName3',
              rank: 3,
              rating: 3,
              username: 'person2',
            },
          ],
          LeaderboardActions.updateType.APPEND,
        ),
      );
    }

    yield put(LeaderboardActions.updateLoadingStatus(false));
  } catch (err) {
    throw err;
  }
}

export function* leaderboardSagas() {
  yield all([takeEvery(LeaderboardActions.Type.GET_LEADERBOARD, getLeaderboard)]);
}
