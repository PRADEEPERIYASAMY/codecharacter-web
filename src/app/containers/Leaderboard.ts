import { RootState } from 'app/reducers';

import { LeaderboardActions } from 'app/actions';
import { Leaderboard } from 'app/components/Leaderboard';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

const mapStateToProps = (rootState: RootState) => {
  return {
    players: rootState.leaderboard.players,
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    getPlayersData: () => dispatch(LeaderboardActions.getPlayersData()),
  };
};

const leaderboardContainer = connect<Leaderboard.StateProps, Leaderboard.DispatchProps, {}>(
  mapStateToProps,
  mapDispatchToProps,
)(Leaderboard);

export default leaderboardContainer;
