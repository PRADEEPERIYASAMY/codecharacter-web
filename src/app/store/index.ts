import { logger } from 'app/middleware';
import { rootReducer } from 'app/reducers';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

export function configureStore(initialState?: object) {
  let middleware = applyMiddleware(logger);

  if (process.env.NODE_ENV !== 'production') {
    middleware = composeWithDevTools(middleware);
  }

  const store = createStore(rootReducer, initialState, middleware);

  if (module.hot) {
    module.hot.accept('app/reducers', () => {
      const nextReducer = require('app/reducers');
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}
