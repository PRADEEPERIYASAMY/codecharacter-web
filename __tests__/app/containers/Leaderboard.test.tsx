import Leaderboard from 'app/containers/Leaderboard';
import { configureStore } from 'app/store';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('Leaderboard Container', () => {
  it('Should render Leaderboard', () => {
    const store = configureStore();
    const wrapper = shallow(<Leaderboard />, {
      context: {
        store,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Should have props', () => {
    const store = configureStore();
    const wrapper = shallow(<Leaderboard />, {
      context: {
        store,
      },
    });
    expect(wrapper.props().players).toEqual(expect.any(Array));
  });
});
