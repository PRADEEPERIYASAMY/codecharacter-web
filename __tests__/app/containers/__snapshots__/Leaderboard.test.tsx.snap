// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard Container Should render Leaderboard 1`] = `
<Leaderboard
  clearLeaderboard={[Function]}
  getLeaderboard={[Function]}
  loading={false}
  players={Array []}
/>
`;
